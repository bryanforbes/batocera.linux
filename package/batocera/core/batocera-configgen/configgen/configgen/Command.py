from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping, Sequence
    from pathlib import Path


class Command:
    def __init__(self, array: Sequence[str | Path], env: Mapping[str, str | Path] = {}):
        self.array = list(array)
        self.env = dict(env)

    def __str__(self):
        strings = [f"{varName}={varValue}" for varName, varValue in self.env.items()]

        strings.extend(str(value) for value in self.array)

        return " ".join(strings)
