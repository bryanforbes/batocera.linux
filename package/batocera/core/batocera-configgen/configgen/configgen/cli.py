from __future__ import annotations

import logging
import time
from collections.abc import Callable
from functools import partial
from pathlib import Path
from typing import Any, TypedDict, TypeVar, Unpack, cast

import click

from . import profiler
from .emulators import run_emulator
from .utils.logger import setup_logging

_FC = TypeVar('_FC', bound=Callable[..., Any] | click.Command)

_logger = logging.getLogger(__name__)


def _player_callback(player: int, ctx: click.Context, option: click.Option, value: int | None) -> int | None:
    current_value: list[PlayerOptions] = ctx.params.setdefault('players', [])

    option_names = (
        f'p{player}hat_count',
        f'p{player}button_count',
        f'p{player}device_path',
        f'p{player}real_name',
        f'p{player}guid',
        f'p{player}index',
    )

    items = [
        (f'p{player}axis_count', value),
        *[(option_name, ctx.params.pop(option_name)) for option_name in option_names],
    ]

    if all(item[1] is not None for item in items):
        p_length = len(f'p{player}')
        current_value.append(
            cast(
                PlayerOptions,
                {
                    'player_number': player,
                    **{option_name[p_length:]: option_value for option_name, option_value in items},
                },
            )
        )
        return value

    if all(item[1] is None for item in items):
        return value

    raise click.UsageError(f'All options that start with p{player} must be specified if any are specified', ctx=ctx)


def _players(max_players: int) -> Callable[[_FC], _FC]:
    def wrap(callable: _FC) -> _FC:
        for n in range(max_players, 0, -1):
            callable = click.option(f'-p{n}nbaxes', f'p{n}axis_count', type=int, callback=partial(_player_callback, n))(
                callable
            )
            callable = click.option(f'-p{n}nbhats', f'p{n}hat_count', type=int)(callable)
            callable = click.option(f'-p{n}nbbuttons', f'p{n}button_count', type=int)(callable)
            callable = click.option(f'-p{n}devicepath', f'p{n}device_path', type=str)(callable)
            callable = click.option(f'-p{n}name', f'p{n}real_name', type=str)(callable)
            callable = click.option(f'-p{n}guid', type=str)(callable)
            callable = click.option(f'-p{n}index', type=int)(callable)

        return callable

    return wrap


class PlayerOptions(TypedDict):
    player_number: int
    index: int
    guid: str
    real_name: str
    device_path: str
    button_count: int
    hat_count: int
    axis_count: int


class Options(TypedDict):
    system: str
    rom: Path
    emulator: str | None
    core: str | None
    netplaymode: str | None
    netplaypass: str | None
    netplayip: str | None
    netplayport: str | None
    netplaysession: str | None
    state_slot: str | None
    state_filename: Path | None
    autosave: str | None
    systemname: str | None
    gameinfoxml: Path
    lightgun: bool
    wheel: bool
    trackball: bool
    spinner: bool
    players: list[PlayerOptions]


@click.command()
@click.option('-system', required=True)
@click.option('-rom', required=True, type=click.Path(path_type=Path))
@click.option('-emulator')
@click.option('-core')
@click.option('-netplaymode')
@click.option('-netplaypass')
@click.option('-netplayip')
@click.option('-netplayport')
@click.option('-netplaysession')
@click.option('-state_slot')
@click.option('-state_filename', type=click.Path(path_type=Path))
@click.option('-autosave')
@click.option('-systemname')
@click.option('-gameinfoxml', default='/dev/null', type=click.Path(path_type=Path))
@click.option('-lightgun', is_flag=True)
@click.option('-wheel', is_flag=True)
@click.option('-trackball', is_flag=True)
@click.option('-spinner', is_flag=True)
@_players(8)
def cli(**kwargs: Unpack[Options]) -> None:
    with setup_logging():
        exit_code = -1

        try:
            exit_code = run_emulator(kwargs)
        except Exception:
            _logger.exception('configgen exception: ')

        profiler.disable(dump=True)

        time.sleep(1)  # this seems to be required so that the gpu memory is restituated and available for es
        _logger.debug('Exiting configgen with status %s', exit_code)

        exit(exit_code)
