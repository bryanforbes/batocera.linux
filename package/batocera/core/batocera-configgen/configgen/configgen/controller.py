from __future__ import annotations

import xml.etree.ElementTree as ET
from collections.abc import Mapping
from dataclasses import dataclass, field, replace
from pathlib import Path
from typing import TYPE_CHECKING, Final, Literal, Self, TypeAlias, TypedDict, Unpack, cast

from .batoceraPaths import BATOCERA_ES_DIR, USER_ES_DIR
from .input import Input, InputDict

if TYPE_CHECKING:
    import argparse

"""Default mapping of Batocera keys to SDL_GAMECONTROLLERCONFIG keys."""
_DEFAULT_SDL_MAPPING: Final = {
    'b': 'a',
    'a': 'b',
    'x': 'y',
    'y': 'x',
    'l2': 'lefttrigger',
    'r2': 'righttrigger',
    'l3': 'leftstick',
    'r3': 'rightstick',
    'pageup': 'leftshoulder',
    'pagedown': 'rightshoulder',
    'start': 'start',
    'select': 'back',
    'up': 'dpup',
    'down': 'dpdown',
    'left': 'dpleft',
    'right': 'dpright',
    'joystick1up': 'lefty',
    'joystick1left': 'leftx',
    'joystick2up': 'righty',
    'joystick2left': 'rightx',
    'hotkey': 'guide'
}


def _key_to_sdl_game_controller_config(keyname: str, name: str, type: str, id: str, value: str) -> str | None:
    """
    Converts a key mapping to the SDL_GAMECONTROLLER format.

    Arguments:
      keyname: (str) SDL_GAMECONTROLLERCONFIG input name.
      name: (str) `es_input.cfg` input name.
      type: (str) 'button', 'hat', or 'axis'
      id: (int) Numeric key id.
      value: (int) Hat value. Only used if type == 'hat' or type == 'axis' and 'joystick' in name.
    Returns:
      (str) SDL_GAMECONTROLLERCONFIG-formatted key mapping string.
    Examples:
      _keyToSdlGameControllerConfig('leftshoulder', 'l1', 'button', 6)
        'leftshoulder:b6'

      _keyToSdlGameControllerConfig('dpleft', 'left', 'hat', 0, 8)
        'dpleft:h0.8'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, -1)
        'lefty:a1'

      _keyToSdlGameControllerConfig('lefty', 'joystick1up', 'axis', 1, 1)
        'lefty:a1~'

      _keyToSdlGameControllerConfig('dpup', 'up', 'axis', 1, -1)
        'dpup:-a1'
    """
    if type == 'button':
        return f'{keyname}:b{id}'
    elif type == 'hat':
        return f'{keyname}:h{id}.{value}'
    elif type == 'axis':
        if 'joystick' in name:
            return f"{keyname}:a{id}{'~' if int(value) > 0 else ''}"
        elif keyname in ('dpup', 'dpdown', 'dpleft', 'dpright'):
            return f"{keyname}:{'-' if int(value) < 0 else '+'}a{id}"
        else:
            return f'{keyname}:a{id}'
    elif type == 'key':
        return None
    else:
        raise ValueError(f'unknown key type: {type!r}')


class _ControllerChanges(TypedDict, total=False):
    guid: str
    player: int | None
    index: int
    real_name: str
    device_path: str | None
    button_count: int | None
    hat_count: int | None
    axis_count: int | None
    physical_device_path: str | None
    physical_id: int | None


@dataclass(slots=True, kw_only=True)
class Controller:
    device_name: str
    type: Literal['keyboard', 'joystick']
    guid: str
    player: int | None
    index: int = -1
    real_name: str = ""
    inputs: InputDict = field(default_factory=dict)
    device_path: str | None = None
    button_count: int | None = None
    hat_count: int | None = None
    axis_count: int | None = None
    physical_device_path: str | None = None
    physical_id: int | None = None

    @property
    def uid_name(self) -> str:
        return self.guid + self.device_name

    def replace(self, /, **changes: Unpack[_ControllerChanges]) -> Self:
        return replace(self, **changes)

    def generate_sdl_game_db_line(self, sdl_mapping: Mapping[str, str] = _DEFAULT_SDL_MAPPING, /) -> str:
        config = [self.guid, self.real_name, 'platform:Linux']

        def add_mapping(input: Input) -> None:
            key_name = sdl_mapping.get(input.name, None)
            if key_name is None:
                return
            sdl_config = _key_to_sdl_game_controller_config(key_name, input.name, input.type, input.id, input.value)
            if sdl_config is not None:
                config.append(sdl_config)

        hotkey_input: Input | None = None
        mapped_button_ids: set[str] = set()
        for input in self.inputs.values():
            if input.name is None:
                continue
            if input.name == 'hotkey':
                hotkey_input = input
                continue
            if input.type == 'button':
                mapped_button_ids.add(input.id)

            add_mapping(input)

        if hotkey_input is not None and hotkey_input.id not in mapped_button_ids:
            add_mapping(hotkey_input)

        config.append('')

        return ','.join(config)

    @classmethod
    def from_element(cls, element: ET.Element, /) -> Self:
        return cls(
            device_name=cast(str, element.get("deviceName")),
            type=cast(Literal['keyboard', 'joystick'], element.get("type")),
            guid=cast(str, element.get("deviceGUID")),
            player=None,
            index=-1,
            inputs=Input.from_parent_element(element)
        )

    @classmethod
    def parse_all(cls) -> list[Self]:
        return [
            cls.from_element(controller)
            for conffile in [BATOCERA_ES_DIR / "es_input.cfg", USER_ES_DIR / 'es_input.cfg'] if conffile.exists()
            for controller in ET.parse(conffile).getroot().findall(".//inputConfig")
        ]

    @classmethod
    def load_all(cls) -> dict[str, Self]:
        return {controller.uid_name: controller for controller in cls.parse_all()}

    @classmethod
    def load_for_players(cls, max_players: int, args: argparse.Namespace, /) -> dict[int, Self]:
        controllers: dict[int, Self] = {}
        all_controllers = cls.load_all()

        for player_number in range(1, max_players + 1):
            controller = cls.find_best_controller_config(all_controllers, args, player_number)
            if controller is not None:
                controllers[player_number] = controller

        return controllers

    @classmethod
    def find_best_controller_config(
        cls, controllers: Mapping[str, Self], args: argparse.Namespace, player_number: int, /,
    ) -> Self | None:
        player_index: int | None = getattr(args, f'p{player_number}index')

        if player_index is None:
            return None

        player_guid: str = getattr(args, f'p{player_number}guid')
        player_name: str = getattr(args, f'p{player_number}name')
        player_device_path: str = getattr(args, f'p{player_number}devicepath')
        player_buttons: int = getattr(args, f'p{player_number}nbbuttons')
        player_hats: int = getattr(args, f'p{player_number}nbhats')
        player_axes: int = getattr(args, f'p{player_number}nbaxes')

        for controller in controllers.values():
            if controller.guid == player_guid and controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.guid == player_guid:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        for controller in controllers.values():
            if controller.device_name == player_name:
                return controller.replace(
                    guid=player_guid,
                    player=player_number,
                    index=player_index,
                    real_name=player_name,
                    device_path=player_device_path,
                    button_count=player_buttons,
                    hat_count=player_hats,
                    axis_count=player_axes,
                )

        return None


ControllerMapping: TypeAlias = Mapping[str, Controller]
ControllerDict: TypeAlias = Mapping[str, Controller]
ControllerPlayerMapping: TypeAlias = Mapping[int, Controller]
ControllerPlayerDict: TypeAlias = dict[int, Controller]


def generate_sdl_game_controller_config(controllers: ControllerPlayerMapping, /) -> str:
    return '\n'.join([controller.generate_sdl_game_db_line() for controller in controllers.values()])

def write_sdl_controller_db(
    controllers: ControllerPlayerMapping, output_file: str | Path = '/tmp/gamecontrollerdb.txt', /,
) -> Path:
    output_file = Path(output_file)
    with output_file.open('w') as text_file:
        text_file.write(generate_sdl_game_controller_config(controllers))
    return output_file
