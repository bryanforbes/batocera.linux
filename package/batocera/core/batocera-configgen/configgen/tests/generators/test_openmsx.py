from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING, Final
from zipfile import ZipFile

import pytest

from configgen.batoceraPaths import CONFIGS, ROMS
from configgen.generators.openmsx.openmsxGenerator import OpenmsxGenerator
from tests.generators.base import GeneratorBaseTest
from tests.mock_controllers import make_player_controller_list

if TYPE_CHECKING:
    from pyfakefs.fake_filesystem import FakeFilesystem
    from syrupy.assertion import SnapshotAssertion

    from configgen.controller import Controller, Controllers
    from configgen.Emulator import Emulator


_DATA_DIR: Final = Path(__file__).parent.parent.parent.parent.parent.parent / 'emulators' / 'openmsx'


class TestOpenmsxGenerator(GeneratorBaseTest):
    @pytest.fixture
    def generator_cls(self) -> type[OpenmsxGenerator]:
        return OpenmsxGenerator

    @pytest.fixture
    def system_name(self) -> str:
        return 'msx1'

    @pytest.fixture
    def emulator(self) -> str:
        return 'openmsx'

    @pytest.fixture(autouse=True)
    def fs(self, fs: FakeFilesystem) -> FakeFilesystem:
        fs.add_real_file(_DATA_DIR / 'settings.xml', target_path='/usr/share/openmsx/settings.xml')
        return fs

    def test_has_internal_mangohud_call(self, generator: OpenmsxGenerator) -> None:  # pyright: ignore
        assert generator.hasInternalMangoHUDCall()

    def test_generate(
        self,
        generator: OpenmsxGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'msx1' / 'rom (foo) [bar].cas',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert (CONFIGS / 'openmsx' / 'share' / 'settings.xml').read_text() == snapshot(name='xml')
        assert (CONFIGS / 'openmsx' / 'share' / 'script.tcl').read_text() == snapshot(name='tcl')

    @pytest.mark.parametrize('system_name', ['msx2', 'msx2+', 'msxturbor', 'colecovision', 'spectravideo'])
    def test_generate_system(
        self,
        generator: OpenmsxGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / mock_system.name / 'rom.cas',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )

    @pytest.mark.parametrize(
        'mock_system_config',
        [
            {'openmsx_loading': 'false'},
            {'hud': 'game'},
            {'hud': ''},
        ],
        ids=str,
    )
    def test_generate_config(
        self,
        generator: OpenmsxGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'msx1' / 'rom.cas',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
        assert (CONFIGS / 'openmsx' / 'share' / 'settings.xml').read_text() == snapshot(name='xml')

    def test_generate_controllers(
        self,
        generator: OpenmsxGenerator,
        mock_system: Emulator,
        generic_xbox_pad: Controller,
        ps3_controller: Controller,
        snapshot: SnapshotAssertion,
    ) -> None:
        generator.generate(
            mock_system,
            ROMS / 'msx1' / 'rom.cas',
            make_player_controller_list(generic_xbox_pad, ps3_controller, generic_xbox_pad),
            {},
            [],
            {},
            {'width': 1920, 'height': 1080},
        )
        assert (CONFIGS / 'openmsx' / 'share' / 'script.tcl').read_text() == snapshot(name='tcl')

    def test_generate_laserdisc(
        self,
        generator: OpenmsxGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'msx1' / 'rom.ogv',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )

    @pytest.mark.parametrize('mock_system_config', [{'openmsx_disk': 'disk'}, {'openmsx_disk': 'hda'}], ids=str)
    def test_generate_disk(
        self,
        generator: OpenmsxGenerator,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        assert (
            generator.generate(
                mock_system,
                ROMS / 'msx1' / 'rom.dsk',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )

    @pytest.mark.parametrize('extension', ['cas', 'dsk', 'ogv'])
    def test_generate_zip(
        self,
        generator: OpenmsxGenerator,
        extension: str,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_dir('/userdata/roms/msx1')
        with ZipFile('/userdata/roms/msx1/rom.zip', 'w') as zip:
            zip.writestr('thing.txt', 'text file')
            zip.writestr(f'rom.{extension}', 'rom')

        assert (
            generator.generate(
                mock_system,
                ROMS / 'msx1' / 'rom.zip',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )

    @pytest.mark.parametrize('extension', ['rom', 'dsk'])
    def test_generate_openmsx_rom(
        self,
        generator: OpenmsxGenerator,
        extension: str,
        fs: FakeFilesystem,
        mock_system: Emulator,
        one_player_controllers: Controllers,
        snapshot: SnapshotAssertion,
    ) -> None:
        fs.create_file('/userdata/roms/msx1/rom.openmsx', contents=f'one.{extension}\ntwo.{extension}\n')

        assert (
            generator.generate(
                mock_system,
                ROMS / 'msx1' / 'rom.openmsx',
                one_player_controllers,
                {},
                [],
                {},
                {'width': 1920, 'height': 1080},
            )
            == snapshot
        )
