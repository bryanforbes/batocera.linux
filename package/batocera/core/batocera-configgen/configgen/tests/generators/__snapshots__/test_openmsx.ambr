# serializer version: 1
# name: TestOpenmsxGenerator.test_generate
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-cassetteplayer',
      '/userdata/roms/msx1/rom (foo) [bar].cas',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate[tcl]
  '''
  filepool add -path /userdata/bios/Machines -types system_rom -position 1
  filepool add -path /userdata/bios/openmsx -types system_rom -position 2
  
  # -= Save state =-
  savestate "rom  "
  
  # -= Screenshots =-
  bind F5 {screenshot [utils::get_next_numbered_filename /userdata/screenshots/openmsx "[guess_title] " ".png"]}
  
  # -= Controller config =-
  plug joyporta joystick1
  dict set joystick1_config LEFT {-axis0 L_hat0}
  dict set joystick1_config RIGHT {+axis0 R_hat0}
  dict set joystick1_config UP {-axis1 U_hat0}
  dict set joystick1_config DOWN {+axis1 D_hat0}
  bind "joy1 button11 down" "toggle_osd_keyboard"
  bind "joy1 button5 up" "set fastforward off"
  bind "joy1 button5 down" "set fastforward on"
  bind "joy1 button12 down" "toggle console"
  bind "joy1 button8 down" "toggle pause"
  bind "joy1 button9 down" "main_menu_toggle"
  bind "joy1 button2 down" "keymatrixdown 6 0x80"
  bind "joy1 button3 down" "keymatrixdown 6 0x40"
  
  '''
# ---
# name: TestOpenmsxGenerator.test_generate[xml]
  '''
  <!DOCTYPE settings SYSTEM 'settings.dtd'>
  <?xml version="1.0" ?>
  <settings>
    <settings>
      <setting id="renderer">SDLGL-PP</setting>
      <setting id="fullscreen">true</setting>
      <setting id="master_volume">100</setting>
      <setting id="save_settings_on_exit">false</setting>
      <setting id="vsync">true</setting>
      <setting id="fullspeedwhenloading">true</setting>
    </settings>
    <bindings>
      <bind key="keyb F6">cycle videosource</bind>
    </bindings>
  </settings>
  '''
# ---
# name: TestOpenmsxGenerator.test_generate_config[{'hud': ''}]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-cassetteplayer',
      '/userdata/roms/msx1/rom.cas',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_config[{'hud': ''}][xml]
  '''
  <!DOCTYPE settings SYSTEM 'settings.dtd'>
  <?xml version="1.0" ?>
  <settings>
    <settings>
      <setting id="renderer">SDLGL-PP</setting>
      <setting id="fullscreen">true</setting>
      <setting id="master_volume">100</setting>
      <setting id="save_settings_on_exit">false</setting>
      <setting id="vsync">true</setting>
      <setting id="fullspeedwhenloading">true</setting>
    </settings>
    <bindings>
      <bind key="keyb F6">cycle videosource</bind>
    </bindings>
  </settings>
  '''
# ---
# name: TestOpenmsxGenerator.test_generate_config[{'hud': 'game'}]
  tuple(
    list([
      'mangohud',
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-cassetteplayer',
      '/userdata/roms/msx1/rom.cas',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_config[{'hud': 'game'}][xml]
  '''
  <!DOCTYPE settings SYSTEM 'settings.dtd'>
  <?xml version="1.0" ?>
  <settings>
    <settings>
      <setting id="renderer">SDLGL-PP</setting>
      <setting id="fullscreen">true</setting>
      <setting id="master_volume">100</setting>
      <setting id="save_settings_on_exit">false</setting>
      <setting id="vsync">true</setting>
      <setting id="fullspeedwhenloading">true</setting>
    </settings>
    <bindings>
      <bind key="keyb F6">cycle videosource</bind>
    </bindings>
  </settings>
  '''
# ---
# name: TestOpenmsxGenerator.test_generate_config[{'openmsx_loading': 'false'}]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-cassetteplayer',
      '/userdata/roms/msx1/rom.cas',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_config[{'openmsx_loading': 'false'}][xml]
  '''
  <!DOCTYPE settings SYSTEM 'settings.dtd'>
  <?xml version="1.0" ?>
  <settings>
    <settings>
      <setting id="renderer">SDLGL-PP</setting>
      <setting id="fullscreen">true</setting>
      <setting id="master_volume">100</setting>
      <setting id="save_settings_on_exit">false</setting>
      <setting id="vsync">true</setting>
      <setting id="fullspeedwhenloading">false</setting>
    </settings>
    <bindings>
      <bind key="keyb F6">cycle videosource</bind>
    </bindings>
  </settings>
  '''
# ---
# name: TestOpenmsxGenerator.test_generate_controllers[tcl]
  '''
  filepool add -path /userdata/bios/Machines -types system_rom -position 1
  filepool add -path /userdata/bios/openmsx -types system_rom -position 2
  
  # -= Save state =-
  savestate "rom"
  
  # -= Screenshots =-
  bind F5 {screenshot [utils::get_next_numbered_filename /userdata/screenshots/openmsx "[guess_title] " ".png"]}
  
  # -= Controller config =-
  plug joyporta joystick1
  dict set joystick1_config LEFT {-axis0 L_hat0}
  dict set joystick1_config RIGHT {+axis0 R_hat0}
  dict set joystick1_config UP {-axis1 U_hat0}
  dict set joystick1_config DOWN {+axis1 D_hat0}
  bind "joy1 button11 down" "toggle_osd_keyboard"
  bind "joy1 button5 up" "set fastforward off"
  bind "joy1 button5 down" "set fastforward on"
  bind "joy1 button12 down" "toggle console"
  bind "joy1 button8 down" "toggle pause"
  bind "joy1 button9 down" "main_menu_toggle"
  bind "joy1 button2 down" "keymatrixdown 6 0x80"
  bind "joy1 button3 down" "keymatrixdown 6 0x40"
  plug joyportb joystick2
  dict set joystick2_config LEFT {-axis0 L_hat0}
  dict set joystick2_config RIGHT {+axis0 R_hat0}
  dict set joystick2_config UP {-axis1 U_hat0}
  dict set joystick2_config DOWN {+axis1 D_hat0}
  bind "joy2 button11 down" "toggle_osd_keyboard"
  bind "joy2 button5 up" "set fastforward off"
  bind "joy2 button5 down" "set fastforward on"
  bind "joy2 button12 down" "toggle console"
  bind "joy2 button8 down" "toggle pause"
  bind "joy2 button9 down" "main_menu_toggle"
  bind "joy2 button2 down" "keymatrixdown 6 0x80"
  bind "joy2 button3 down" "keymatrixdown 6 0x40"
  
  '''
# ---
# name: TestOpenmsxGenerator.test_generate_disk[{'openmsx_disk': 'disk'}]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-diska',
      '/userdata/roms/msx1/rom.dsk',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_disk[{'openmsx_disk': 'hda'}]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-hda',
      '/userdata/roms/msx1/rom.dsk',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_laserdisc
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Pioneer_PX-7',
      '-laserdisc',
      '/userdata/roms/msx1/rom.ogv',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_openmsx_rom[dsk]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-diska',
      '/userdata/roms/msx1/one.dsk',
      '-diskb',
      '/userdata/roms/msx1/two.dsk',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_openmsx_rom[rom]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-carta',
      '/userdata/roms/msx1/one.rom',
      '-cartb',
      '/userdata/roms/msx1/two.rom',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_system[colecovision]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'ColecoVision_SGM',
      '-cassetteplayer',
      '/userdata/roms/colecovision/rom.cas',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_system[msx2+]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2+_JP',
      '-cassetteplayer',
      '/userdata/roms/msx2+/rom.cas',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_system[msx2]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-cassetteplayer',
      '/userdata/roms/msx2/rom.cas',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_system[msxturbor]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSXturboR_with_IDE',
      '-cassetteplayer',
      '/userdata/roms/msxturbor/rom.cas',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_system[spectravideo]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Spectravideo_SVI-328',
      '-cassetteplayer',
      '/userdata/roms/spectravideo/rom.cas',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_zip[cas]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-cassetteplayer',
      '/userdata/roms/msx1/rom.zip',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_zip[dsk]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Boosted_MSX2_EN',
      '-diska',
      '/userdata/roms/msx1/rom.zip',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_generate_zip[ogv]
  tuple(
    list([
      '/usr/bin/openmsx',
      '-machine',
      'Pioneer_PX-7',
      '-laserdisc',
      '/userdata/roms/msx1/rom.zip',
      '-script',
      '/userdata/system/configs/openmsx/share/script.tcl',
    ]),
    dict({
      'XDG_DATA_HOME': '/userdata/system/configs/openmsx/share',
    }),
  )
# ---
# name: TestOpenmsxGenerator.test_get_hotkeys_context
  dict({
    'keys': dict({
      'exit': list([
        'KEY_LEFTALT',
        'KEY_F4',
      ]),
      'restore_state': 'KEY_F6',
    }),
    'name': 'openmsx',
  })
# ---
